### PR TITLE
Make ParameterSource a proper Python 3 Enum subclass.

### LINKED ISSUE #1530: Make ParameterSource an Enum
`core.ParameterSource` was implemented as fake enum for compatibility with Python 2. Replace this with a simple `enum.Enum` now that we support Python >= 3.6. Can remove the `ParameterSource.validate` method too, it was just for ensuring the enum values were used.

### PR DESCRIPTION
**Pull request by alanbato**
This implements the change proposed in  #1530.

I tried to make as few changes as possible, that's why I went and implemented a custom `__eq__()` to support the current use case inside the library (mostly inside tests)

Any comments and feedback are welcome :)

fixes #1530

### CODE REVIEW

File: tests/test_context.py [hunk 1]
```diff
@@ -323,5 +323,5 @@ def cli(ctx, option):
 
 
 def test_validate_parameter_source():
-    with pytest.raises(ValueError):
-        ParameterSource.validate("NOT_A_VALID_PARAMETER_SOURCE")
+    with pytest.raises(KeyError):
+        ParameterSource["NOT_A_VALID_PARAMETER_SOURCE"]
```
Comment (ThiefMaster) [inline tests/test_context.py:327]: Probably this test can go away. Testing basic stdlib functionality doesn't make much sense IMO

Comment (alanbato) [reply tests/test_context.py:327]: I agree, I wasn't sure if I should remove the test but I'm happy to do so after your suggestion

File: src/click/core.py [hunk 2]
```diff
@@ -126,36 +127,27 @@ def sort_key(item):
...
     """This is an enum that indicates the source of a command line parameter.
 
     The enum has one of the following values: COMMANDLINE,
     ENVIRONMENT, DEFAULT, DEFAULT_MAP.  The DEFAULT indicates that the
-    default value in the decorator was used.  This class should be
-    converted to an enum when Python 2 support is dropped.
+    default value in the decorator was used.
+
+    .. versionchanged:: 8
+        This class is now a proper Python3 Enum
```
Comment (jab) [inline src/click/core.py:138]:
```suggestion
    .. versionchanged:: 8.0
        Made this an :class:`~enum.Enum`
        and dropped the now-superfluous ``validate`` method. 
```

Comment (jab) [reply src/click/core.py:138]: (Important to note when public API is removed)

**Comment by jab** [comment]
> I tried to make as few changes as possible, that's why I went and implemented a custom __eq__() to support the current use case inside the library

A reasonable first cut, for sure. But @davidism, not sure if that's what you had in mind. It seems to me that changing the usage sites to do direct enum comparison (rather than having to keep the custom `__eq__` implementation here) would be more in the spirit of this change. WDYT?

**Pull request merged**